lib/ansible/module_utils/common/arg_spec.py

```diff
@@ -36,13 +42,29 @@
 class ArgumentSpecValidator():
     """Argument spec validation class"""
 
-    def __init__(self, argument_spec, parameters):
+    def __init__(self, argument_spec, parameters,
+                 mutually_exclusive=None,
+                 required_together=None,
+                 required_one_of=None,
+                 required_if=None,
+                 required_by=None,
+                 _kind='module',
+                 _name=None,
```

Comment:
I'm on the fence here with `_kind` and `_name`.  I think I'd much rather the caller be responsible for this.  But there are tradeoffs.

A possible solution would be to use `{{ kind }}` and `{{ name }}` in the string formatting in here:

```
self._add_error("Unsupported parameters for ({{name}}) {{kind}}: {0}. Supported parameters include: {1}."
```

and then allow the caller to do something like:

```
self.fail_json(msg=self.validator.error_messages[0].format(kind='module', name=self._name))
```

Reply:
I didn't really like this much either. It doesn't seem like information that is relevant to the validator class and is only used in warnings. I mainly did this to appease the existing unit tests which set `AnsibleModule._name` and look for it in the output string.

Reply:
I'll play around with the format string option. That seems like a good way to keep the context in the caller.

Reply:
This does work, and I like that it keeps those bits of information out of the validator class. But it does mean the caller could end up with `Unsupported parameters for ({name}) {kind}:` in their error messages if they don't replace those values. That seems a bit awkward.

Reply:
I would suggest that this validator class not depend on knowing those caller specifics at all, nor put the responsibility of filling in templated error messages on the caller. How about, instead, the `error_messages` attribute be generic lines of text without that data, and we offer a new `error_json` attribute that the caller could parse to generate it's own error messages with any specifics it has?

Reply:
I've been thinking the same thing.  Ideally it would be great if the validator raised exceptions, instead of passing back formatted errors, but I get the reasons why it is like it is currently.

Reply:
I explored what it would take to do this, and it's not as straight forward as I'd hoped it would be. Much of the validation actually happens in `parameters.py`, so being able to categorize errors into a json structure in the validator class in `arg_spec.py` would be difficult. It could be done, just would need some slight re-engineering. While I think that would be a more elegant and clean solution to this, I don't think time is going to allow for it and we should probably just stick with what Sam has here now.

FWIW, I think it would be possible to implement the structured data solution later, while also keeping the current solution, so we could provide a better way forward for handling errors and at the same time not break anything currently using the templated message behavior.

Reply:
After talking to @Shrews, I have an idea how to add an interface for raising exceptions vs. returning the formatted messages. But I don't think it will be done in time for release.

Reply:
Unsupported parameters are the only type of failure that requires more context than can be derived from the argument spec and parameters. Maybe we could raise an exception just for unsupported parameters? Or provide a public `unsupported_parameters` attribute the caller needs to handle? Neither are ideal since it makes the interface more complicated, but it would eliminate this problem.